sklearn/decomposition/tests/test_kernel_pca.py

```diff
@@ -272,3 +276,100 @@ def test_nested_circles():
     # The data is perfectly linearly separable in that space
     train_score = Perceptron(max_iter=5).fit(X_kpca, y).score(X_kpca, y)
     assert_equal(train_score, 1.0)
+
+
+def test_errors_and_warnings():
+    """Tests that bad kernels raise error and warnings"""
+
+    solvers = ['dense', 'arpack']
+    solvers_except_arpack = ['dense']
+
+    # First create an identity transformer class
+    # ------------------------------------------
+    class IdentityKernelTransformer(BaseEstimator, TransformerMixin):
+        """We will use this transformer so that the passed kernel matrix is
+        not centered when kPCA is fit"""
+
+        def __init__(self):
+            pass
+
+        def fit(self, K, y=None):
+            return self
+
+        def transform(self, K, y=None, copy=True):
+            return K
+
+    # Significant imaginary parts: error
+    # ----------------------------------
+    # As of today it seems that the kernel matrix is always cast as a float
+    # whatever the method (precomputed or callable).
+    # The following test therefore fails ("did not raise").
+    K = [[5, 0],
+         [0, 6 * 1e-5j]]
+    # for solver in solvers:
+    #     kpca = KernelPCA(kernel=kernel_getter, eigen_solver=solver,
+    #                      fit_inverse_transform=False)
+    #     kpca._centerer = IdentityKernelTransformer()
+    #     K = kpca._get_kernel(K)
+    #     with pytest.raises(ValueError):
+    #         # note: we can not test 'fit' because _centerer would be replaced
+    #         kpca._fit_transform(K)
+    #
+    # For safety concerning future evolutions the corresponding code is left in
+    # KernelPCA, and we test it directly by calling the inner method here:
+    with pytest.raises(ValueError):
+        check_kernel_eigenvalues((K[0][0], K[1][1]))
+
+    # All negative eigenvalues: error
+    # -------------------------------
+    K = [[-5, 0],
+         [0, -6e-5]]
+    # check that the inner method works
+    with pytest.raises(ValueError):
+        check_kernel_eigenvalues((K[0][0], K[1][1]))
+
+    for solver in solvers:
+        kpca = KernelPCA(kernel="precomputed", eigen_solver=solver,
+                         fit_inverse_transform=False)
+        kpca._centerer = IdentityKernelTransformer()
+        K = kpca._get_kernel(K)
+        with pytest.raises(ValueError):
+            # note: we can not test 'fit' because _centerer would be replaced
+            kpca._fit_transform(K)
+
+    # Significant negative eigenvalue: warning
+    # ----------------------------------------
+    K = [[5, 0],
+         [0, -6e-5]]
+    # check that the inner method works
+    assert_warns(KernelWarning,
```

Comment:
instead of `assert_warns` please use `with pytest.warns(KernelWarning, match=warning_mesg):` instead

Reply:
ok